cpu/efm32/include/periph_cpu.h

```diff
@@ -245,6 +198,183 @@ typedef enum {
+ */
+#define ADC_MODE_OSR(mode)       ((mode & 0xff0000) >> 16)
+
+#if defined(_IADC_CFG_DIGAVG_MASK)
+/**
+ * @brief   Internal macro to extract over-sampling rate
+ */
+#define ADC_MODE_AVG(mode)       ((mode & 0x00ff00) >> 8)
+#endif
+
+/**
+ * @brief   Internal macro to extract output resolution
+ */
+#define ADC_MODE_RES(mode)       ((mode & 0x0000ff) >> 0)
+
+/**
+ * @brief   Possible ADC resolution settings
+ * @{
+ */
+#define HAVE_ADC_RES_T
+#if defined(_IADC_CFG_DIGAVG_MASK)
+typedef enum {
+    ADC_RES_6BIT  = ADC_MODE(iadcCfgOsrHighSpeed2x, iadcDigitalAverage1, 6),
+    ADC_RES_8BIT  = ADC_MODE(iadcCfgOsrHighSpeed2x, iadcDigitalAverage1, 8),
+    ADC_RES_10BIT = ADC_MODE(iadcCfgOsrHighSpeed2x, iadcDigitalAverage1, 10),
+    ADC_RES_12BIT = ADC_MODE(iadcCfgOsrHighSpeed2x, iadcDigitalAverage1, 12),
+    ADC_RES_14BIT = ADC_MODE(iadcCfgOsrHighSpeed8x, iadcDigitalAverage1, 14),
+    ADC_RES_16BIT = ADC_MODE(iadcCfgOsrHighSpeed16x, iadcDigitalAverage2, 16),
+} adc_res_t;
+#else
+typedef enum {
+    ADC_RES_6BIT  = ADC_MODE(iadcCfgOsrHighSpeed2x, 6),
+    ADC_RES_8BIT  = ADC_MODE(iadcCfgOsrHighSpeed2x, 8),
+    ADC_RES_10BIT = ADC_MODE(iadcCfgOsrHighSpeed2x, 10),
+    ADC_RES_12BIT = ADC_MODE(iadcCfgOsrHighSpeed2x, 12),
+    ADC_RES_14BIT = ADC_MODE(iadcCfgOsrHighSpeed8x, 14),
+    ADC_RES_16BIT = ADC_MODE(iadcCfgOsrHighSpeed32x, 16),
+} adc_res_t;
+#endif
+
+/**
+ * @brief   ADC device configuration
+ */
+typedef struct {
+    /**
+     * IADC device configuration
+     */
+    IADC_TypeDef *dev;
+
+    /**
+     * CMU gate for the IADC device
+     */
+    CMU_Clock_TypeDef cmu;
+
+    /**
+     * Voltage reference to use
+     */
+    IADC_CfgReference_t reference;
+
+    /**
+     * Voltage of the reference in mV
+     *
+     * @note  Required internally for offset correction.
+     */
+    uint32_t reference_mV;
+
+    /**
+     * Ampilfication of the analog input signal
+     *
+     * @note  The maximum input voltage is
+     *        \ref adc_conf_t.gain * \ref adc_conf_t.reference_mV
+     */
+    IADC_CfgAnalogGain_t gain;
+
+    /**
+     * Available resoltions
+     *
+     * @note  Resolutions made available to the applications have to be
+     *        specified during \ref adc_init. This will configure the IADC
+     *        accordingly and allows for quick \ref adc_sample calls.
+     */
+    adc_res_t available_res[IADC0_CONFIGNUM];
+} adc_conf_t;
+
+/**
+ * @brief   ADC channel configuration
+ */
+typedef struct {
+    /**
+     * \ref adc_conf_t device index
+     */
+    uint8_t dev;
+
+    /**
+     * Positive analog input
+     */
+    gpio_t input_pos;
+
+    /**
+     * Negative analog input.
+     * Can be set to \ref GPIO_UNDEF for single-ended ADC lines.
+     *
+     * @note  For differential inputs make sure that
+     *        \ref adc_chan_conf_t.input_pos is an even pin number and
+     *        \ref adc_chan_conf_t.input_neg is an odd pin number or the other
+     *        way around.
```

Comment:
But the pins don't need to be consecutive, right?

Reply:
No they don't. This is the internal implementation of the ADC routing:

![Screenshot from 2022-11-21 11-58-00](https://user-images.githubusercontent.com/6105784/203033680-51983991-494a-4cfa-883d-eea255a4bc0b.png)

The swap of the odd and even line is carried out by hardware automatically without any software interaction.